File: src/Collega.API/Controllers/IdeaAssistController.cs
using Collega.API.Contracts.Ai;
using Collega.Application.Ai;
using Collega.Domain.Enums;
using Microsoft.AspNetCore.Authorization;
using Microsoft.AspNetCore.Mvc;

namespace Collega.API.Controllers;

/// <summary>
/// AI-assisted idea drafting (SPEC/30-Contracts.md → "AI Idea Assist Contracts"). Thin, like every
/// controller here: the scope gate, id re-validation, turn caps, metering and audit all live in
/// <see cref="IIdeaAssistService"/>.
/// </summary>
/// <remarks>
/// This endpoint <b>never creates, updates, or deletes an idea</b> (rule 23). It returns draft
/// suggestions that seed the create form, which is submitted separately through
/// <c>POST /boards/{boardId}/ideas</c> and validated there as normal. That separation is the reason
/// no model output is ever authorization-bearing, and it must not be "optimized" away.
/// </remarks>
[ApiController]
[Route("boards/{boardId:guid}/idea-assist")]
public sealed class IdeaAssistController : ControllerBase
{
    private readonly IIdeaAssistService _ideaAssist;

    public IdeaAssistController(IIdeaAssistService ideaAssist)
    {
        _ideaAssist = ideaAssist;
    }

    /// <summary>Advances the drafting conversation by one turn.</summary>
    /// <remarks>
    /// <c>503</c> means the assistant is unavailable — unconfigured, provider down, or the daily token
    /// budget is exhausted. The three are indistinguishable on purpose; clients degrade to the
    /// scripted brainstorm rather than surfacing an error.
    /// </remarks>
    [Authorize]
    [HttpPost("turns")]
    [ProducesResponseType(typeof(IdeaAssistTurnResponse), StatusCodes.Status200OK)]
    [ProducesResponseType(typeof(ValidationProblemDetails), StatusCodes.Status400BadRequest)]
    [ProducesResponseType(typeof(ProblemDetails), StatusCodes.Status401Unauthorized)]
    [ProducesResponseType(typeof(ProblemDetails), StatusCodes.Status403Forbidden)]
    [ProducesResponseType(typeof(ProblemDetails), StatusCodes.Status404NotFound)]
    [ProducesResponseType(typeof(ProblemDetails), StatusCodes.Status503ServiceUnavailable)]
    public async Task<IActionResult> Continue(
        Guid boardId,
        [FromBody] IdeaAssistTurnRequestContract request,
        CancellationToken cancellationToken)
    {
        var result = await _ideaAssist.ContinueAsync(
            new IdeaAssistTurnRequest(
                boardId,
                request.Transcript.Select(t => new IdeaAssistTurn(t.Role, t.Text)).ToList(),
                ToDraft(request.Draft)),
            cancellationToken);

        return Ok(new IdeaAssistTurnResponse(
            result.InScope,
            result.ConversationClosed,
            result.NextQuestion,
            IdeaDraftContract.From(result.Draft),
            result.TurnsRemaining));
    }

    private static IdeaDraft? ToDraft(IdeaDraftContract? contract)
    {
        if (contract is null)
        {
            return null;
        }

        // A priority the client cannot parse is dropped rather than rejected — same forgiveness the
        // contract gives unknown option ids, and for the same reason.
        var priority = Enum.TryParse<Priority>(contract.Priority, ignoreCase: true, out var parsed)
            ? parsed
            : (Priority?)null;

        return new IdeaDraft(
            contract.Title,
            contract.Description,
            contract.IdeaTypeId,
            contract.BusinessImpactId,
            priority);
    }
}

/// <summary>
/// Deployment-level availability of AI idea assist (rule 32a). Its own controller because the route is
/// not org-scoped and, unlike <see cref="OrganizationAiAssistSettingsController"/>, is open to any
/// authenticated user.
/// </summary>
[ApiController]
[Route("ai-assist")]
public sealed class AiAssistAvailabilityController : ControllerBase
{
    private readonly IIdeaAssistService _ideaAssist;

    public AiAssistAvailabilityController(IIdeaAssistService ideaAssist)
    {
        _ideaAssist = ideaAssist;
    }

    /// <summary>
    /// Whether the client should open the drafting chat or go straight to the create form.
    /// </summary>
    /// <remarks>
    /// <para><b>Any authenticated user</b>, deliberately: creating ideas is a <c>User</c>-role activity,
    /// so the admin-only settings endpoint could not answer this question for the people who need it.
    /// Safe to widen because the response is one boolean — no key material, no org configuration, no
    /// usage figures — and it makes no provider call, so it is neither metered nor rate limited.</para>
    /// <para>It also never says <i>why</i>. Unconfigured, provider down and budget exhausted collapse
    /// into one <c>false</c>, exactly as they collapse into one <c>503</c> on a turn (rule 31); a
    /// client that could tell them apart would start treating them differently.</para>
    /// </remarks>
    [Authorize]
    [HttpGet("availability")]
    [ProducesResponseType(typeof(AiAssistAvailabilityResponse), StatusCodes.Status200OK)]
    [ProducesResponseType(typeof(ProblemDetails), StatusCodes.Status401Unauthorized)]
    public async Task<IActionResult> GetAvailability(CancellationToken cancellationToken) =>
        Ok(new AiAssistAvailabilityResponse(await _ideaAssist.IsAvailableAsync(cancellationToken)));
}

/// <summary>
/// Site-Admin management of the idea-assist prompt (SPEC/20-feature-ai-idea-assist.md rules 34–38).
/// </summary>
/// <remarks>
/// <para>Deployment configuration, not organization content — the same scope as the API key in rule 29,
/// which is why the route is not org-scoped and why this does not go through View As.</para>
///
/// <para>Role gating is enforced <b>again</b> in <see cref="IAiPromptService"/> rather than relying on
/// the attribute here. The service check reads the effective role, so it also refuses a Site Admin who
/// is currently acting as someone else — something the attribute alone would let through.</para>
/// </remarks>
[ApiController]
[Route("ai-assist/prompt")]
[Authorize(Roles = "SiteAdmin")]
public sealed class AiPromptController : ControllerBase
{
    private readonly IAiPromptService _prompts;

    public AiPromptController(IAiPromptService prompts)
    {
        _prompts = prompts;
    }

    /// <summary>The active prompt plus its version history. Returns the built-in default when none is published.</summary>
    [HttpGet]
    [ProducesResponseType(typeof(AiPromptResponse), StatusCodes.Status200OK)]
    [ProducesResponseType(typeof(ProblemDetails), StatusCodes.Status401Unauthorized)]
    [ProducesResponseType(typeof(ProblemDetails), StatusCodes.Status403Forbidden)]
    public async Task<IActionResult> Get(CancellationToken cancellationToken) =>
        Ok(ToResponse(await _prompts.GetAsync(cancellationToken)));

    /// <summary>Publishes a new version and makes it active.</summary>
    [HttpPut]
    [ProducesResponseType(typeof(AiPromptResponse), StatusCodes.Status200OK)]
    [ProducesResponseType(typeof(ValidationProblemDetails), StatusCodes.Status400BadRequest)]
    [ProducesResponseType(typeof(ProblemDetails), StatusCodes.Status401Unauthorized)]
    [ProducesResponseType(typeof(ProblemDetails), StatusCodes.Status403Forbidden)]
    public async Task<IActionResult> Publish(
        [FromBody] PublishAiPromptRequest request,
        CancellationToken cancellationToken)
    {
        var settings = await _prompts.PublishAsync(
            new PublishAiPromptCommand(request.Body, request.OutOfScopeRedirect, request.ConversationClosedRedirect),
            cancellationToken);

        return Ok(ToResponse(settings));
    }

    /// <summary>Republishes an earlier version as a new one, keeping history append-only.</summary>
    [HttpPost("versions/{version:int}/restore")]
    [ProducesResponseType(typeof(AiPromptResponse), StatusCodes.Status200OK)]
    [ProducesResponseType(typeof(ProblemDetails), StatusCodes.Status401Unauthorized)]
    [ProducesResponseType(typeof(ProblemDetails), StatusCodes.Status403Forbidden)]
    [ProducesResponseType(typeof(ProblemDetails), StatusCodes.Status404NotFound)]
    public async Task<IActionResult> Restore(int version, CancellationToken cancellationToken) =>
        Ok(ToResponse(await _prompts.RestoreAsync(version, cancellationToken)));

    /// <summary>Returns the deployment to the built-in default by standing every version down.</summary>
    [HttpPost("reset")]
    [ProducesResponseType(typeof(AiPromptResponse), StatusCodes.Status200OK)]
    [ProducesResponseType(typeof(ProblemDetails), StatusCodes.Status401Unauthorized)]
    [ProducesResponseType(typeof(ProblemDetails), StatusCodes.Status403Forbidden)]
    public async Task<IActionResult> Reset(CancellationToken cancellationToken) =>
        Ok(ToResponse(await _prompts.ResetToDefaultAsync(cancellationToken)));

    /// <summary>
    /// Runs the advisory safety probes against a draft (rule 37). Never publishes, and a failing probe
    /// never blocks a later publish.
    /// </summary>
    /// <remarks>
    /// <c>503</c> here means the probes could not run — unconfigured, provider down, or out of daily
    /// budget. It deliberately does not report "refused", because reporting an outage as a passed probe
    /// is the one wrong answer this surface can give.
    /// </remarks>
    [HttpPost("probe")]
    [ProducesResponseType(typeof(AiPromptProbeResponse), StatusCodes.Status200OK)]
    [ProducesResponseType(typeof(ValidationProblemDetails), StatusCodes.Status400BadRequest)]
    [ProducesResponseType(typeof(ProblemDetails), StatusCodes.Status401Unauthorized)]
    [ProducesResponseType(typeof(ProblemDetails), StatusCodes.Status403Forbidden)]
    [ProducesResponseType(typeof(ProblemDetails), StatusCodes.Status503ServiceUnavailable)]
    public async Task<IActionResult> Probe(
        [FromBody] ProbeAiPromptRequest request,
        CancellationToken cancellationToken)
    {
        var report = await _prompts.ProbeAsync(request.Body, cancellationToken);

        return Ok(new AiPromptProbeResponse(
            report.Probes
                .Select(p => new AiPromptProbeItemResponse(p.Id, p.Prompt, p.Refused, p.ExpectedRefused))
                .ToList(),
            report.RefusedCount,
            report.TotalCount));
    }

    private static AiPromptResponse ToResponse(AiPromptSettings settings) => new(
        settings.Body,
        settings.OutOfScopeRedirect,
        settings.ConversationClosedRedirect,
        settings.Version,
        settings.IsBuiltInDefault,
        settings.Versions
            .Select(v => new AiPromptVersionResponse(
                v.Version, v.CreatedAtUtc, v.CreatedByUserId, v.CreatedByDisplayName, v.IsActive))
            .ToList());
}

/// <summary>
/// The organization's AI assist configuration — the scope statement an Org Admin tunes (D-SCOPE).
/// Its own controller so the route sits under <c>organizations/{organizationId}</c>.
/// </summary>
[ApiController]
[Route("organizations/{organizationId:guid}/ai-assist")]
public sealed class OrganizationAiAssistSettingsController : ControllerBase
{
    private readonly IIdeaAssistService _ideaAssist;

    public OrganizationAiAssistSettingsController(IIdeaAssistService ideaAssist)
    {
        _ideaAssist = ideaAssist;
    }

    /// <summary>Reads the configuration. Never returns a key or any part of one.</summary>
    [Authorize(Roles = "OrgAdmin,SiteAdmin")]
    [HttpGet("settings")]
    [ProducesResponseType(typeof(AiAssistSettingsResponse), StatusCodes.Status200OK)]
    [ProducesResponseType(typeof(ProblemDetails), StatusCodes.Status401Unauthorized)]
    [ProducesResponseType(typeof(ProblemDetails), StatusCodes.Status403Forbidden)]
    [ProducesResponseType(typeof(ProblemDetails), StatusCodes.Status404NotFound)]
    public async Task<IActionResult> GetSettings(Guid organizationId, CancellationToken cancellationToken)
    {
        var settings = await _ideaAssist.GetSettingsAsync(organizationId, cancellationToken);
        return Ok(new AiAssistSettingsResponse(settings.AiAssistAvailable, settings.ScopeStatement));
    }

    /// <summary>
    /// Sets or clears the scope statement. Takes effect on the next turn — in-flight conversations
    /// are not retroactively re-scoped.
    /// </summary>
    [Authorize(Roles = "OrgAdmin,SiteAdmin")]
    [HttpPut("settings")]
    [ProducesResponseType(typeof(AiAssistSettingsResponse), StatusCodes.Status200OK)]
    [ProducesResponseType(typeof(ValidationProblemDetails), StatusCodes.Status400BadRequest)]
    [ProducesResponseType(typeof(ProblemDetails), StatusCodes.Status401Unauthorized)]
    [ProducesResponseType(typeof(ProblemDetails), StatusCodes.Status403Forbidden)]
    [ProducesResponseType(typeof(ProblemDetails), StatusCodes.Status404NotFound)]
    public async Task<IActionResult> UpdateSettings(
        Guid organizationId,
        [FromBody] UpdateAiAssistSettingsRequest request,
        CancellationToken cancellationToken)
    {
        var settings = await _ideaAssist.SetScopeStatementAsync(organizationId, request.ScopeStatement, cancellationToken);
        return Ok(new AiAssistSettingsResponse(settings.AiAssistAvailable, settings.ScopeStatement));
    }
}

File: apps/api/src/ai-assist/idea-assist.controller.ts
import type {
  AiAssistSettings,
  IdeaAssistTurn,
  IdeaAssistTurnResult,
  IdeaDraft,
} from '@collega/application/ai'
import {
  IDEA_ASSIST_ASSISTANT_ROLE,
  IDEA_ASSIST_USER_ROLE,
  IdeaAssistService,
  TRANSCRIPT_ENTRY_MAX_LENGTH,
} from '@collega/application/ai'
import { Priority, Role } from '@collega/domain/enums'
import { DESCRIPTION_MAX_LENGTH, TITLE_MAX_LENGTH } from '@collega/domain/ideas'
import { ORGANIZATION_AI_SCOPE_STATEMENT_MAX_LENGTH } from '@collega/domain/organizations'
import { Body, Controller, Get, Param, Post, Put, UseGuards } from '@nestjs/common'
import { AuthGuard } from '../auth/auth.guard.js'
import { Roles } from '../auth/roles.decorator.js'
import { RolesGuard } from '../auth/roles.guard.js'
import {
  displayName,
  type FieldRules,
  RequestValidationError,
  validateFields,
} from '../common/errors/request-validation.error.js'
import { optional, optionalGuid } from '../common/request-values.js'
import { UuidParamPipe } from '../common/uuid-param.pipe.js'

/** The draft fields, on both the request and the response (D-PREFILL). */
type IdeaDraftBody = {
  title?: string | null
  description?: string | null
  ideaTypeId?: string | null
  businessImpactId?: string | null
  priority?: string | null
}

/** `POST /boards/{boardId}/idea-assist/turns` request body. */
type IdeaAssistTurnBody = {
  transcript?: unknown
  draft?: IdeaDraftBody | null
}

/** `PUT /organizations/{organizationId}/ai-assist/settings` request body. */
type UpdateAiAssistSettingsBody = { scopeStatement?: string | null }

/** `GET /ai-assist/availability` - one boolean, and deliberately nothing else. */
type AiAssistAvailabilityResponse = { available: boolean }

const TRANSCRIPT_ROLES = [IDEA_ASSIST_USER_ROLE, IDEA_ASSIST_ASSISTANT_ROLE] as const

/**
 * Reads the transcript array and transcribes the per-entry attributes
 * `IdeaAssistTurnContract` carries - `[RequiredField]` and `[AllowedValues]` on `role`,
 * `[RequiredField]` and `[MaxLengthField(4000)]` on `text`.
 *
 * The entry COUNT, the last-entry-is-user rule and the trimming are `IdeaAssistService`'s, not
 * this function's: they are contract rules the service already owns and re-checks for every
 * caller, and duplicating them here would put the same constraint in two places with two
 * different envelopes.
 *
 * An omitted `transcript` reads as empty rather than missing, which is not an oversight: the .NET
 * property was initialised to `new()` and System.Text.Json leaves an initialiser alone for a key
 * that is not in the JSON, so `[RequiredField]` never fired and the empty list reached the service
 * - which answers "At least one message is required." in the Application envelope. An explicit
 * `null` DID overwrite the initialiser and fail the attribute, which is the branch below.
 */
function readTranscript(value: unknown): readonly IdeaAssistTurn[] {
  if (value === undefined) {
    return []
  }
  if (value === null) {
    throw new RequestValidationError({ transcript: [`${displayName('transcript')} is required.`] })
  }
  if (!Array.isArray(value)) {
    // The one place this reader refuses rather than defaults, following `stringList`'s reasoning
    // in `common/request-values.ts`: reading a mistyped transcript as absent would ACCEPT the
    // request and silently discard what the caller asked about.
    throw new RequestValidationError({ transcript: [`${displayName('transcript')} is invalid.`] })
  }

  const rules: Record<string, FieldRules> = {}
  const entries: IdeaAssistTurn[] = []

  for (const [index, item] of value.entries()) {
    // A non-object entry is read as one with no properties, so it fails `role`/`text` presence
    // rather than throwing on a property read - the same forgiveness `validateFields` gives a
    // mistyped scalar, and for the same reason: a 500 on a malformed body is the worse answer.
    const entry = (typeof item === 'object' && item !== null ? item : {}) as Partial<IdeaAssistTurn>
    // Keyed by the whole path and named by the property alone - ASP.NET keyed `Transcript[0].Role`
    // (camelCased) but built the message from the property's own display name, so the entry reads
    // "Role is required.", never "transcript[0].role is required.".
    rules[`transcript[${index}].role`] = {
      value: entry.role,
      required: true,
      allowedValues: TRANSCRIPT_ROLES,
      displayName: 'Role',
    }
    rules[`transcript[${index}].text`] = {
      value: entry.text,
      required: true,
      maxLength: TRANSCRIPT_ENTRY_MAX_LENGTH,
      displayName: 'Text',
    }
    entries.push({
      role: typeof entry.role === 'string' ? entry.role : '',
      text: typeof entry.text === 'string' ? entry.text : '',
    })
  }

  validateFields(rules)
  return entries
}

/**
 * `Enum.TryParse<Priority>(ignoreCase: true)`: a priority the client cannot spell is DROPPED
 * rather than rejected, the same forgiveness the contract gives an unknown option id, and for the
 * same reason - a stale value is a client that fell behind, not an attack.
 */
function readPriority(value: unknown): Priority | null {
  const text = optional(value)
  if (text === null) {
    return null
  }
  const lower = text.trim().toLowerCase()
  return Object.values(Priority).find((p) => p.toLowerCase() === lower) ?? null
}

/**
 * The `draft` the client echoes back so the model revises rather than restates. Unknown or
 * inactive ids are discarded by `IdeaAssistService` rather than rejected here, so the GUID
 * coercion's empty-GUID fallback resolves against nothing and falls out on its own.
 */
function readDraft(body: IdeaDraftBody | null | undefined): IdeaDraft | null {
  if (body === null || body === undefined) {
    return null
  }

  validateFields({
    title: { value: body.title, maxLength: TITLE_MAX_LENGTH },
    description: { value: body.description, maxLength: DESCRIPTION_MAX_LENGTH },
  })

  return {
    title: optional(body.title),
    description: optional(body.description),
    ideaTypeId: optionalGuid(body.ideaTypeId),
    businessImpactId: optionalGuid(body.businessImpactId),
    priority: readPriority(body.priority),
  }
}

/**
 * AI-assisted idea drafting, availability, and the per-organization scope statement
 * (`SPEC/30-Contracts.md` "AI Idea Assist Contracts"; behaviour in
 * `SPEC/20-feature-ai-idea-assist.md`).
 *
 * **None of these routes ever creates, updates or deletes an idea** (rule 23). The turn endpoint
 * returns draft suggestions that seed the create form, which is submitted separately through
 * `POST /boards/{boardId}/ideas` and validated there as normal. That separation is why no model
 * output is ever authorization-bearing, and it must not be "optimized" away.
 *
 * **Nothing the client sends steers the assistant.** There is no prompt, no system instructions, no
 * model name, no retrieved context, no scope statement and no organization id on the way in - the
 * server assembles all of it from the caller's own identity, which is what keeps a client from
 * reaching another tenant's data.
 *
 * `@Controller()` carries no prefix on purpose - the three routes sit under three different roots,
 * so any prefix would move at least two of them. See `boards.controller.ts` for the convention.
 */
@Controller()
export class IdeaAssistController {
  constructor(private readonly ideaAssist: IdeaAssistService) {}

  /**
   * Advances the drafting conversation by one turn.
   *
   * `503` means the assistant is unavailable - unconfigured, provider down, or the deployment's
   * daily token budget exhausted. The three are indistinguishable on purpose: all of them mean
   * "keep working without the assistant", and a client that could tell them apart would start
   * treating them differently. `AiAssistUnavailableError` carries no cause for the same reason.
   *
   * `AuthGuard` only - the role gate is `IdeaAssistService.requireDraftingMember`, which refuses
   * Read Only and refuses a Site Admin acting as themselves, each with its own recorded message.
   */
  @Post('boards/:boardId/idea-assist/turns')
  @UseGuards(AuthGuard)
  async continueTurn(
    @Param('boardId', UuidParamPipe) boardId: string,
    @Body() body: IdeaAssistTurnBody,
  ): Promise<IdeaAssistTurnResult> {
    return this.ideaAssist.continueTurn({
      boardId,
      transcript: readTranscript(body.transcript),
      draft: readDraft(body.draft),
    })
  }

  /**
   * Whether the client should open the drafting chat or go straight to the create form (rule 32a).
   *
   * **Any authenticated user**, deliberately: creating ideas is a `User`-role activity, so the
   * admin-only settings route below could not answer this question for the people who need it.
   * Safe to widen because the answer is one boolean - no key material, no organization
   * configuration, no usage figures - and it makes no provider call, so it is neither metered nor
   * rate limited.
   *
   * A snapshot, not a subscription: the budget can be exhausted between this call and the next
   * turn, so a client must still handle the turn endpoint's `503`.
   */
  @Get('ai-assist/availability')
  @UseGuards(AuthGuard)
  async availability(): Promise<AiAssistAvailabilityResponse> {
    return { available: await this.ideaAssist.isAvailable() }
  }

  /** Reads the organization's configuration. Never returns a key, nor any part of one (rule 28). */
  @Get('organizations/:organizationId/ai-assist/settings')
  @UseGuards(AuthGuard, RolesGuard)
  @Roles(Role.OrgAdmin, Role.SiteAdmin)
  async settings(
    @Param('organizationId', UuidParamPipe) organizationId: string,
  ): Promise<AiAssistSettings> {
    return this.ideaAssist.getSettings(organizationId)
  }

  /**
   * Sets or clears the scope statement - the free-text narrowing of what the assistant will
   * discuss. Null or empty clears it, leaving the organization's active Idea Types as the only
   * boundary. Takes effect on the next turn; in-flight conversations are not retroactively
   * re-scoped.
   */
  @Put('organizations/:organizationId/ai-assist/settings')
  @UseGuards(AuthGuard, RolesGuard)
  @Roles(Role.OrgAdmin, Role.SiteAdmin)
  async updateSettings(
    @Param('organizationId', UuidParamPipe) organizationId: string,
    @Body() body: UpdateAiAssistSettingsBody,
  ): Promise<AiAssistSettings> {
    validateFields({
      scopeStatement: {
        value: body.scopeStatement,
        maxLength: ORGANIZATION_AI_SCOPE_STATEMENT_MAX_LENGTH,
      },
    })

    return this.ideaAssist.setScopeStatement(organizationId, optional(body.scopeStatement))
  }
}
